Enhance CMake compiler check on Linux

cgmb opened the pull request:
Somehow I missed the specified compiler version in the docs. Maybe I forgot about it by the time I got to the platform-specific requirements. Maybe I just can't read. Not sure what the problem was, but it would be nice for CMake to catch that sort of mistake.

This commit adds a warning for compilers that don't meet the project's minimum requirements. There is already an explicit check for GNU, but this will catch Clang versions that are too low as well as any non-GNU, non-Clang compilers.

The check is Linux-only as I have only built Filament on Linux and I'm not sure if this check would apply to Windows or OSX. The message is warning rather than an error to minimize the impact of false-positives caused by differing compiler names or version numbers. In particular, forks of clang may be compatible but have different names or numbers.

I know the contribution guidelines suggest I should open an issue before submitting a PR, but I figured this was a small enough change that it was probably unnecessary. I'm guessing the process was set up to avoid wasted development effort on rejected designs, but if there's some other reason I should be opening an issue first, I'd be happy to go back and do that.

romainguy commented on CMakeLists.txt:
```diff
@@ -61,6 +61,26 @@ set(TOOLS ${CMAKE_SOURCE_DIR}/tools)
+    endif()
+endif()
+
 if ("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
```
This test here becomes useless with the tests you added.

romainguy commented on CMakeLists.txt:
```diff
@@ -61,6 +61,26 @@ set(TOOLS ${CMAKE_SOURCE_DIR}/tools)
 # ==================================================================================================
 # Compiler check
 # ==================================================================================================
+if (LINUX)
```
No need to check only on Linux.

cgmb commented:
I've removed the GNU check, applied my checks to all platforms, rebased against master and updated the commit message to match. I have only tested on Linux, so I'm crossing my fingers there's no surprises elsewhere.

I'm not entirely certain that WARNING is the best message level, but I figured it was better to be cautious. If you ever want to make it a hard error, just change the message level to SEND_ERROR.

romainguy commented on CMakeLists.txt:
```diff
@@ -61,11 +61,22 @@ set(TOOLS ${CMAKE_SOURCE_DIR}/tools)
+
+if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    if (CMAKE_C_COMPILER_VERSION VERSION_LESS MIN_CLANG_VERSION)
+        message(WARNING "Detected C compiler Clang ${CMAKE_C_COMPILER_VERSION} < ${MIN_CLANG_VERSION}")
```
The old code was using `FATAL_ERROR`, you should use the same level. `WARNING` won't stop the build but since it won't build… might as well error out early.

You should also `return()`.

romainguy commented on CMakeLists.txt:
```diff
@@ -61,11 +61,22 @@ set(TOOLS ${CMAKE_SOURCE_DIR}/tools)
-    return()
+set(MIN_CLANG_VERSION "5.0")
+
+if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
```
`clang` on macOS is called `AppleClang`, you need to check for that compiler ID as well.

cgmb commented:
I believe I've addressed the issue with OSX, though the CI will have to test it for me. I've also changed the message from a warning to an error, though I don't bail out early. I think that's preferable, but will defer to your judgment if you feel strongly about it.

romainguy commented:
You're right, the return is not useful. In this case I think fatal error makes sense because if we proceed further we might just get meaningless errors due to the use of the wrong compiler. I see this check as a fatal precondition. For other errors, sure send would be useful

romainguy commented:
Your change runs just fine on macOS by the way, thanks for doing this!
